### ISSUE TITLE
Use complete sets automatically when they are small and needed

### ISSUE THREAD

**Issue by jarioksa** [gavinsimpson/permute#10]
Should it be useful to switch to `shuffleSet(..., control = how(complete = TRUE)` when `nperm`  is a high proportion of all permutations or in general when the number of all permutations is low? Now automatic setting only happens when the user asks for too many permutations:
```R
> numPerms(7)
[1] 5040
> sum(duplicated(shuffleSet(7, 5041)))
'nperm' > set of all permutations; Resetting 'nperm'.
[1] 0
> sum(duplicated(shuffleSet(7, 5040)))
[1] 1893
## duplicates also with smaller samples
> sum(duplicated(shuffleSet(7, 999)))
[1] 104
```
Please note also the borderline case of resetting the number of permutations to `numPerms(object) - 1`:
```R
> dim(shuffleSet(7, 5040))
[1] 5040    7
> dim(shuffleSet(7, 5041))
'nperm' > set of all permutations; Resetting 'nperm'.
[1] 5039    7
```
Both of these problems will be avoided if `shuffleSet`  resets `how(complete = TRUE)`  if `numPerms(object)`  is low (say, 50000).

**Comment by gavinsimpson** [comment]
This can currently be configured to happen automagically by setting `minperm` to an appropriate value; the default is `99`. We could change `minperm` to `minperm = 50000` (would need to check the logic in the function that does the checking, as now `minperm` > `maxperm` by default.

Or we could leave `minperm` alone and add a different heuristic based on `nperm` and `numPerms()` and allow a user-configurable value of 20% (?) of total possible; in this case if requested permutations was 20% or more of the total possible we'd kick in to complete enumeration. I could call this `proportion = 0.2`?

Thoughts? Either way I'd like this to be user-configurable rather than hard-coded as in your fork currently.

Re the borderline case (if you were pointing this out as a potential issue --- ignore this if you weren't) returns 1 permutation fewer than the total as it doesn't return the observed permutation of the data. You'd need `observed = TRUE` in the `how()` call for that:

```r
> dim(shuffleSet(7, 5041, how(observed = TRUE)))
'nperm' > set of all permutations; Resetting 'nperm'.
[1] 5040    7
```

**Comment by jarioksa** [comment]
My idea was that setting `complete = TRUE` is an automatic and simple precaution in `check()`. I think there is no need to have complicated parameters to `check` because user has all control in `how()`: `check` is only a dumb precaution against naive misuse, but you can always take full control despite `check` conventions (leg, shoot, can do).

**Comment by gavinsimpson** [comment]
Right, but I would prefer to set the `factorial(8)` limit as an argument rather than hard-code this limit in `check()`. Also, this is really what `minperm` was designed for although I didn't really consider the fact that even with relatively large numbers of permutations (~50K ish) we'd still run the risk of an appreciable number of duplicate permutations.

One problem with doing any of this is that it is slow for order `factorial(8)` and above:

```r
> system.time(shuffleSet(8, factorial(8), how(minperm = factorial(8) + 1)))
Set of permutations < 'minperm'. Generating entire set.
   user  system elapsed 
  3.911   0.002   3.930
```

which is an order of magnitude slower than `factorial(7)`

```r
> system.time(x <- dim(shuffleSet(7, factorial(7), control = how(minperm = factorial(8) + 1))))
Set of permutations < 'minperm'. Generating entire set.
   user  system elapsed 
  0.266   0.000   0.267
```

Doug Bates supplied the current version of the code used to generate all permutations of a vector, which was faster than my original. Doug also supplied a much faster version which used functions from the Eigen linear algebra c++ library. I didn't want to put this into **permute** though as it would ratchet up the dependencies and make the package more complex.

**Comment by jarioksa** [comment]
I was checking "typical" `nperm = 999` cases. Here results from simple permutations:
```R
> sum(duplicated(shuffleSet(7, 999)))
[1] 91
> sum(duplicated(shuffleSet(8, 999)))
[1] 16
> sum(duplicated(shuffleSet(9, 999)))
[1] 1
```
For 7! permutations you get 10% duplicates, for 8! still 1% duplicates, and actually only 9! permutations are safe, but 9! = 362880 which is too high in general. So I settled for a limit of 8! = 40320 which does not create too large internal matrices but avoids excessive number of duplicates. The exact point is arbitrary, but I think the balance is quite nice here. You can always use `check = FALSE` and explicit `control = how(…)`. I just think that users may not notice this, and there is your foot again waiting to be shot at.

**Comment by gavinsimpson** [comment]
OK, I'm with you now. If I implement this change as you have it in your fork, I will probably have to look into a faster algorithm, probably in C, to try to bring down the compute time for the `factorial(8)` case in `allPerms()`. 4 seconds to just generate the permutations, most of which get thrown away, is a bit much. Guess I'd better go and relearn some C :-)

Do you want to do a PR from your fork and I'll merge it into this version?

**Comment by jarioksa** [comment]
My bad -- I didn't see `minperm` which indeed seems to do what I suggested with this change. Only problem with that (at least with the version of **pemute** I happen to have in this laptop) is that it gives wrong information about the number of requested permutations.
```R
> p <- shuffleSet(8, 999, control=how(minperm=40321))
Set of permutations < 'minperm'. Generating entire set.
> attr(p, "control")

Permutation Design:

Blocks:
  Defined by: none

Plots:
  Defined by: none

Within Plots:
  Permutation type: free

Permutation details:
  Number of permutations requested: 40320
  Max. number of permutations allowed: 40320
  Evaluate all permutations?: Yes.  Activation limit: 40321
> dim(p)
[1] 999   8
```
This generated 999 permutations like requested, but claims that 40320 was requested.

**Comment by jarioksa** [comment]
I'll see about the PR tomorrow. Just finished looking at Henry IV, part 2 (BBC series) and going to bed now. Better sleep before pull requests.

I agree 4 secs is long, but duplicating 100 cases of slow analysis sometimes takes even longer.

**Comment by gavinsimpson** [comment]
I'll look at the `minperm` issue. It's a clash between how you set up the design and what `shuffleSet()` ended up doing. Should be a simple fix.

**Comment by jarioksa** [comment]
After night's sleep, here some thoughts.

There are two general principles:

1. It is reasonable to expect that there are no duplicated permutations and the original order (that we put back among observations when evaluating *P*-values) is not among permutations. We cannot guarantee neither of these expectations, but number of possible simple permutations is so large that we are fairly safe even in small data sets. Even with  *N*=10 we have 3.6 million possible permutations, and non-unique permutations are only prominent for really small data sets (*N* < 8). However, the number of possible permutations can be really low in several restricted schemes.

2. `check` is a dumb safeguard for a casual user. It should catch worst cases, but it will leave some responsibility to the user. Moreover, user must be able to over-rule `check` decisions. I also think that `check` could be loud and alert users to take precautions even when `check` does not make them automatically. 

With these principles I suggest `check` should work so that:

- If the number of possible permutations (`num.pos`) is very low, we sample from complete enumeration to get unique permutations without duplicates and without original order. I suggested limit 8! = 40320 for `num.pos`, but this may indeed be too high (although `allPerms` for this number only took 0.5 sec in my desktop). I suggest 7! = 5040 as a general limit. For *N*=8 and `complete=FALSE` we may be satisfied that users get 900-some unique permutations when they request 999. Perhaps a warning of duplicates could be issued when `num.pos` is lowish (50000 or less?) so that users can set `complete = TRUE`.

- If the users *knows* the number of possible permutations and request that number (or `numpos - 1` because we do not generate the original order), they should get all possible permutations. Now they get 63.2% with duplicates instead of complete enumeration.

- Finally, if the users request too many permutations, we reset `nperm` to `num.pos - 1` and give a complete enumeration. This already happens now. 

I can prepare a PR along these lines. There is a separate issue of API, but I leave that to another comment.

**Comment by jarioksa** [comment]
This comment is about API. My user scenario is that I use `shuffleSet()`  to generate a permutation matrix, and there I need `how()` to define the permutation design. I find the `how` especially cryptic and sometimes tedious to work: you must set many things to change one. Here comments on some `how` arguments:

- You can set `nperm` both in `shuffleSet()` and in `how()`, and it seems that `shuffleSet()` has the priority. A problem with this is that although you can set `nperm`  in `how()` you do not then know what is the actual `nperm` used when `how()` is invoked. It could be argued that `nperm` is not an attribute of sampling design (= `how()`), but it is only a property of sampling realization from that design (= `shuffleSet()`). In any cases, the usefulness of `nperm` in `how()` is low because you do not know if that value is ever used.

- I find it confusing that I read the documentation so that you should be able to set complete enumeration with three alternative ways: `complete = TRUE`, `minperm = numPerms()`  and `make = TRUE`. Currently some of these are slightly buggy. `complete = TRUE` is OK, but you must set `minperm`  to larger than number of possible permutations, and I have no idea how to make `make  TRUE` work.

- Documentation says that currently `maxperm` is ignored, but this is not true: you must set it. I find it bizarre that you need to set `maxperm`  to be able to set `complete = TRUE`  or `minperm`.

- You must set both `maxperm` and `minperm` to a larger value than `num.pos`

To generate the complete set  of 40319 unique simple permutations (excluding observed order), I can use these:

```R
shuffleSet(8, control=how(nperm=40319, complete=TRUE, maxperm=40321))
shuffleSet(8, control=how(nperm=40319, minperm=40321, maxperm=40321))
```

**Comment by jarioksa** [comment]
With borderline cases I meant this:
```R
> dim(shuffleSet(7, 5041))
'nperm' > set of all permutations; Resetting 'nperm'.
[1] 5039    7
> dim(shuffleSet(7, 5040))
[1] 5040    7
> dim(shuffleSet(7, 5039))
[1] 5039    7
```
The maximum number of permutation is 7! = 5040, but we do not return the original order and hence we should return 5039 permutations. What happened above:
- requested 5041: received complete enumeration of 5039 unique permutations
- requested 5040: received 5040 non-unique permutations with duplicates
- requested 5039: received 5039 non-unique permutation with duplicates

**Comment by gavinsimpson** [comment]
Thanks for the additional comments and thoughts @jarioksa. `check()` got a little neglected during the various changes that went into **permute** prior to us getting it working within **vegan**. I'll take a look at the API for check and ponder how this should all come together.

The reasoning behind having `nperm` in `how()` was that you need it to work out if you need to hand-hold the user. `shuffleSet()` allows you to not set `nset` if the value is set in `how()`, and if you forget both you 199 permutations which is the default. `nset` in `shuffleSet()` should have the priority if and only if `nset <= nperm`, where `nperm` is what is set *after* the design has been checked. As you showed, there is a buglet in that in the case where `nperm` is set to some new value based on `check()` and `nset` is **less than** this value, we still retain and show `nperm` as the number of **requested** permutations, which is wrong and should be fixed.

`make` is a bit of a fudge, IIRC, to work round some inflexibility in the way `check()` and `allPerms()` worked together. `make` isn't intended as something a normal user should be fiddling with. I could make this clearer in the help. It's really for people programming with **permute** if you want to check the permutation design but not generate the entire set of permutations (bc the heuristics might trigger & without `make` there is no way to stop that). I could even rename this argument to `.make` and move it to the end of the argument list to further downplay that it shouldn't actually be used.

Are you sure you need to set `minperm` *and* `complete = TRUE` to get that to work? I don't see that (copied from above):

```r
> system.time(shuffleSet(8, factorial(8), how(minperm = factorial(8) + 1)))
Set of permutations < 'minperm'. Generating entire set.
   user  system elapsed 
  3.911   0.002   3.930
```

The general overview as I envisaged it was:

 * If `complete = TRUE`, this trumps everything and we generate all permutation *unless* this would generate `maxperm` permutations as a guard against doing something that would take a long time to return. I should probably increase the default for `maxperm` and especially if we do as you suggest and generate all permutations in more settings.

    `complete = TRUE` is the way for the user to signal that they *really* want all permutations, regardless if the heuristics in `check()` would trigger or not.

 * `minperm` is there to set the lower limit at which heuristics kick in in `check()` *in regard to whether complete enumeration is desirable*; I set this low, especially so in light of the examples you show above. I would consider increasing this to 999 now.

    I could have `minperm` trigger if the number of permutations is equal to or less than `minperm`; which would avoid the issue of having to set it to a value 1 larger than the number of permutations ot get it to work.

 * `maxperm` is there to stop things blowing up, but is user-configurable so you can allow things to blow up if you have a lot of RAM and want to run the permutations in parallel.

 * `make` is for developers that might need to check a design without letting `check()` trigger stuff. `check` doesn't have this argument as I envisaged it being read from the `how()` design. This seems like a mistake now.

    Perhaps a solution there is to give `check()` an argument `make` and remove this from `how()`. Hence as a developer you would need to indicate this when calling `check()` rather than embed it in the permutation design. That would help separate this argument and the user.

I can fix the API along these lines (or others if you have suggestions?)

**Comment by gavinsimpson** [comment]
@jarioksa Re your thoughts on how this should work, I agree and thanks for articulating them. I'd be happy to merge a PR that implemented those 3 bullets.

**Comment by jarioksa** [comment]
I think I have now implemented those three bulleted points of triggering complete enumerations. We have complete enumerations if (1) nperm > num.pos, or (2) if nperm == num.pos, or (3) nperm >= minperm. Still need to check these.

Some details that need checking:

* the max number of permutations is one lower if we do not request observed order. I have taken this into account in some points, but code needs review.

* `getNperm` does not return `integer` but `double`: we need `EPS` in comparisons! I think `getNperm` must return `double` because the result can overflow integer.

* Usually we should not `setNperm` because *request*  did not change (we did not fulfill the request), and usually we also return `nperm` permutations. The only case where `nperm` should be reset is when it was too high and we return `num.pos - !getObserved(control)`. This is, naturally, a policy question but if we call it *requested* number, it should be such. I suggest that in general it should be number of delivered permutations or `min(nperm, num.pos - !getObserved)`. 

I won't make this a PR yet. I need to review the code, in particular for `getObserved()`.

**Comment by jarioksa** [comment]
So this is not going to make it to 0.9?

**Comment by gavinsimpson** [comment]
Bugger... I thought I'd checked all of this. I can release 0.9-1 or 0.10-0 soon if this wasn't included. I was getting multiple people hitting bugs with 0.8-4 that were fixed here. Sorry if I have overlooked this - it wasn't intentional al all :-(

**Comment by gavinsimpson** [comment]
Some of this has certainly made it into 0.9-0 - you supplied a PR on Aug 30th including much of what we discuss in #11

**Comment by jarioksa** [comment]
Gav, I think I did not see all your merges. There were some issues in **vegan**`::nesteddisc()`, but when I check against the comments on those kluges, it seems that all were solved in this release. I make **vegan** dependent on **permute**`_0.9-0` and fix the code.
